apis/appmesh/v1beta2/virtualrouter_types.go

```diff
@@ -0,0 +1,367 @@
+	Prefix *string `json:"prefix,omitempty"`
+	// An object that represents the range of values to match on.
+	// +optional
+	Range *MatchRange `json:"range,omitempty"`
+	// The value sent by the client must include the specified characters.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=255
+	// +optional
+	Regex *string `json:"regex,omitempty"`
+	// The value sent by the client must end with the specified characters.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=255
+	// +optional
+	Suffix *string `json:"suffix,omitempty"`
+}
+
+// HttpRouteHeader refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HttpRouteHeader.html
+type HttpRouteHeader struct {
+	// A name for the HTTP header in the client request that will be matched on.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=50
+	Name string `json:"name"`
+	// The HeaderMatchMethod object.
+	// +optional
+	Match *HeaderMatchMethod `json:"match,omitempty"`
+	// Specify True to match anything except the match criteria. The default value is False.
+	// +optional
+	Invert *bool `json:"invert,omitempty"`
+}
+
+// HttpRouteMatch refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HttpRouteMatch.html
+type HttpRouteMatch struct {
+	// An object that represents the client request headers to match on.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=10
+	// +optional
+	Headers []HttpRouteHeader `json:"headers,omitempty"`
+	// The client request method to match on.
+	// +optional
+	Method *string `json:"method,omitempty"`
+	// Specifies the path to match requests with
+	Prefix string `json:"prefix"`
+	// The client request scheme to match on
+	// +kubebuilder:validation:Enum=http;https
+	// +optional
+	Scheme *string `json:"scheme,omitempty"`
+}
+
+// HttpRouteAction refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HttpRouteAction.html
+type HttpRouteAction struct {
+	// An object that represents the targets that traffic is routed to when a request matches the route.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=10
+	WeightedTargets []WeightedTarget `json:"weightedTargets"`
+}
+
+// +kubebuilder:validation:Enum=server-error;gateway-error;client-error;stream-error
+type HTTPRetryEvent string
+
+// +kubebuilder:validation:Enum=connection-error
+type TCPRetryEvent string
+
+// +kubebuilder:validation:Enum=cancelled;deadline-exceeded;internal;resource-exhausted;unavailable
+type GRPCRetryEvent string
+
+// HttpRetryPolicy refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HttpRetryPolicy.html
+type HttpRetryPolicy struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=25
+	// +optional
+	HTTPRetryEvents []HTTPRetryEvent `json:"httpRetryEvents,omitempty"`
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=1
```

Comment:
Not entirely sure what an enforced single-item array gives us that wouldn't be better represented as a scalar element, but OK.

Reply:
Hi jay, i got it from the APIDatastructure: https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_HttpRetryPolicy.html
I think it's good to align with api(which uses a single-item array), since it will help us align with api changes in the future. (e.g. more tcp events)